mkosi/machine.py

```diff
@@ -0,0 +1,104 @@
+
+from __future__ import annotations
+
+import contextlib
+import enum
+import signal
+from subprocess import PIPE, CompletedProcess
+import sys
+from typing import Any, List, Optional, Sequence, Union
+
+import pexpect # type: ignore
+
+from . import (
+    MKOSI_COMMANDS_SUDO,
+    build_stuff,
+    bump_image_version,
+    check_native,
+    check_output,
+    check_root,
+    init_namespace,
+    load_args,
+    needs_build,
+    parse_args,
+    prepend_to_environ_path,
+    print_output_size,
+    run_qemu_cmdline,
+    run_shell_cmdline,
+    save_manifest,
+    suppress_stacktrace,
+    unlink_output,
+    virt_name,
+    run_command_image
+)
+from .backend import die, run, _FILE
+
+
+class Machine:
+    def __init__(self, args: Optional[Sequence[str]] = None) -> None:
+        # Remains None until image is built and booted, then receives pexpect process
+        self._serial: Optional[pexpect.spawn] = None
+        self.stack = contextlib.ExitStack()
+        self.args = load_args(parse_args(args)["default"])
+        self.exit_code: int = 1
+
+    @property
+    def serial(self) -> pexpect.spawnu:
+        if self._serial is None:
+            raise ValueError("Trying to access serial console before machine boot or after machine shutdown.")
+
+        return self._serial
+
+
+    def __enter__(self) -> Machine:
+        with contextlib.ExitStack() as stack:
+            prepend_to_environ_path(self.args.extra_search_paths)
+
+            if self.args.verb in MKOSI_COMMANDS_SUDO:
+                check_root()
+                unlink_output(self.args)
+
+            if self.args.verb == "build":
+                check_output(self.args)
+
+            if needs_build(self.args):
+                check_root()
+                check_native(self.args)
+                init_namespace(self.args)
+                manifest = build_stuff(self.args)
+
+                if self.args.auto_bump:
+                    bump_image_version(self.args)
+
+                save_manifest(self.args, manifest)
+
+                print_output_size(self.args)
+
+            if self.args.verb in ("shell", "boot"):
+                cmdline = run_shell_cmdline(self.args)
+            elif self.args.verb == "qemu":
+                # We must keep the temporary file opened at run_qemu_cmdline accessible, hence the context stack.
+                cmdline = stack.enter_context(run_qemu_cmdline(self.args))
+            else:
+                die("No valid verb was entered.")
+
+            cmdline = [str(x) for x in cmdline]
+            cmdline = " ".join(cmdline)
+
+            self._serial = pexpect.spawnu(cmdline, logfile=None, timeout=240)
+            self._serial.expect("#")
+            self.stack = stack.pop_all()
+
+        return self
+
+
+    def machine_run(self, commands: List[str], stdout: _FILE = sys.stdout, stderr: _FILE = sys.stderr) -> CompletedProcess:
+        return run_command_image(self.args, commands, PIPE, PIPE)
```

Comment:
Yeah, I didn't realize I wasn't using the parameters lol. But what if someone wants to direct run_command_image()'s output to the Pipe or Console, how would they do it without the arguments from machine_run()?

Unless maybe we remove machine_run() completely and start calling run_command_image() directly?

Reply:
I'm also using the stdout/stderr arguments in the test_machine.py file in order to assert that the run function is running correctly, since the pexpect.expect() function wasn't able to find the output of subprocess.run(). So. now, I redirect output to PIPE and then assert it with .stdout.

Reply:
I agree that it's useful to have commands log to stdout/stderr for debugging, but if we implement that by passing sys.stdout and sys.stderr to run_command_image(), we'll break the corresponding assertions since the stdout and stderr fields of the CompletedProcess will be empty.

It would also be quite painful for users to have to change the stdout/stderr arguments for every call to machine_run() when debugging. Instead, we could have a `debug` field as part of `Machine` that can be toggled on/off by the user. When enabled, after `run_command_image` finishes, we print the stdout and stderr fields of CompletedProcess.